common/beerocks/bcl/include/bcl/network/socket_refactorings.h

```diff
@@ -0,0 +1,795 @@
+     */
+    FileDescriptorImpl(const FileDescriptorImpl &) = delete;
+
+    /**
+     * Delete assignment operator to avoid having multiple copies of the file descriptor
+     */
+    FileDescriptorImpl &operator=(const FileDescriptorImpl &) = delete;
+
+    /**
+     * A move constructor allows the resources owned by an rvalue object to be moved into an
+     * lvalue without creating its copy.
+     */
+    FileDescriptorImpl(FileDescriptorImpl &&obj)
+    {
+        this->m_fd = obj.m_fd;
+        obj.m_fd   = FileDescriptor::invalid_descriptor;
+    }
+
+    /**
+     * Close socket on destructor (if it is still open)
+     */
+    ~FileDescriptorImpl() override { close(); }
+
+    /**
+     * @brief Gets file descriptor.
+     *
+     * File descriptor can be used:
+     * - in select(), poll() or epoll() to wait for events on this descriptor.
+     * - in send() family of functions to send data through the socket connection.
+     * - in recv() family of functions to receive data from the socket connection.
+     *
+     * @return File descriptor value.
+     */
+    int fd() override { return m_fd; }
+
+    /**
+     * @brief Closes file descriptor.
+     *
+     * If valid, closes file descriptor and then invalidates it.
+     *
+     * @return True on success and false otherwise (for example, if it was already closed)
+     */
+    bool close() override
+    {
+        printf("close(%d)\n", m_fd);
+        if (FileDescriptor::invalid_descriptor == m_fd) {
+            return false;
+        }
+
+        int rc = ::close(m_fd);
+        m_fd   = FileDescriptor::invalid_descriptor;
+
+        return (0 == rc);
+    }
+
+private:
+    int m_fd;
+};
+
+/**
+ * Wrapper class around sockadr (structure describing a generic socket address)
+ */
+class SockAddr {
+public:
+    virtual ~SockAddr()                             = default;
+    virtual const struct sockaddr *sockaddr() const = 0;
+    virtual const socklen_t &length() const         = 0;
+
+    struct sockaddr *sockaddr()
+    {
+        return const_cast<struct sockaddr *>(const_cast<const SockAddr *>(this)->sockaddr());
+    }
+    socklen_t &length()
+    {
+        return const_cast<socklen_t &>(const_cast<const SockAddr *>(this)->length());
+    }
+};
+
+/**
+ * Wrapper class around msghdr (structure describing messages sent by `sendmsg` and received by
+ * `recvmsg`)
+ */
+class MsgHdr {
+public:
+    virtual ~MsgHdr()                           = default;
+    virtual const struct msghdr *msghdr() const = 0;
+
+    struct msghdr *msghdr()
+    {
+        return const_cast<struct msghdr *>(const_cast<const MsgHdr *>(this)->msghdr());
+    }
+};
+
+/**
+ * Any object implementing the Socket interface is a FileDescriptor too.
+ * This way we give meaning to the action of "closing the socket".
+ */
+class Socket : public FileDescriptor {
+    virtual int setsockopt(int level, int optname, void *optval, socklen_t optlen) = 0;

```

Comment:
There is a mix of C and C++ in this refactoring.
I think we should strive to go C++ with all its benefits.
* `void*` (and the inevitable accompanied `length`)  **must** not be used within prplMesh unless inside wrappers that calls 3rd party interfaces that asks for it. users of interfaces must have nothing to do with void*, ever.
* we should make simple things simple for our users:
```suggestion
    virtual int setsockopt(PrplMeshSocketOptions) = 0;
    // and of course define a clean, simple and safe PrplMeshSocketOptions class.
    // this class encapsulate _only_ the parameters that we need and keep everything else opaque to the user and 
    // correct to the 3rd party that needs it
```

Reply:
Probably separate functions for each setsockopt variant that we actually use is more appropriate.